Content.Server/Language/ChatListener.LanguageCoupling.cs

```diff
@@ -0,0 +1,136 @@
+            // it made the most sense for the language listeners to know about
+            // SayListener et al, rather than the other listeners to explicitly
+            // know about Language.
+
+            if (!args.Chat.GetData<bool>(ChatDataSay.IsSpoken))
+                return;
+
+            if (!args.Chat.TryGetData<LanguagePrototype>(ChatDataLanguage.Language, out var language))
+                return;
+
+            var name = args.Chat.Source;
+            var message = args.RecipientData.GetData<string>(ChatRecipientDataSay.Message) ?? args.Chat.Message;
+
+            if (args.RecipientData.GetData<bool>(ChatRecipientDataLanguage.IsUnderstood))
+            {
+                if (args.RecipientData.GetData<bool>(ChatRecipientDataLanguage.IsSpeakingSameLanguage))
+                    return;
+
+                if (args.Chat.ClaimedBy == typeof(SayListenerSystem))
+                {
+                    args.RecipientData.SetData(ChatRecipientDataSay.WrappedMessage, Loc.GetString("chat-manager-entity-say-language-wrap-message",
+                        ("entityName", name),
+                        ("message", message),
+                        ("language", language.Name)));
+                }
+                else if (args.Chat.ClaimedBy == typeof(WhisperListenerSystem))
+                {
+                    args.RecipientData.SetData(ChatRecipientDataSay.WrappedMessage, Loc.GetString("chat-manager-entity-whisper-language-wrap-message",
+                        ("entityName", name),
+                        ("message", message),
+                        ("language", language.Name)));
+                }
+                else if (args.Chat.ClaimedBy == typeof(RadioListenerSystem))
+                {
+                    var channel = args.RecipientData.GetData<RadioChannelPrototype>(ChatRecipientDataRadio.SharedRadioChannel);
+
+                    if (channel == null)
+                    {
+                        // Receiving only the whisper part.
+                        args.RecipientData.SetData(ChatRecipientDataSay.WrappedMessage, Loc.GetString("chat-manager-entity-radio-language-wrap-message",
+                            ("entityName", name),
+                            ("message", message),
+                            ("language", language.Name)));
+                    }
+                    else
+                    {
+                        args.RecipientData.SetData(ChatRecipientDataSay.WrappedMessage, Loc.GetString("chat-radio-language-message-wrap",
+                            ("color", channel.Color),
+                            ("channel", $"\\[{channel.LocalizedName}\\]"),
+                            ("name", name),
+                            ("message", message),
+                            ("language", language.Name)));
+                    }
+                }
+
+                return;
+            }
+
+            var unknownLanguage = Loc.GetString("chat-manager-unknown-language");
```

Comment:
Don't GetString this, pass in the identifier directly and use `LOC()` inside the translation.